google_reauth/reauth.py

```diff
@@ -0,0 +1,276 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A module that provides functions for handling rapt authentication."""
+
+import base64
+import getpass
+import json
+import logging
+import sys
+
+from six.moves import urllib
+
+from pyu2f import errors as u2ferrors
+from pyu2f import model
+from pyu2f.convenience import authenticator
+
+from google_reauth import reauth_errors
+
+
+REAUTH_API = 'https://reauth.googleapis.com/v2/sessions'
+REAUTH_SCOPE = 'https://www.googleapis.com/auth/accounts.reauth'
+REAUTH_ORIGIN = 'https://accounts.google.com'
+
+
+logger = logging.getLogger(__name__)
+
+
+def HandleErrors(msg):
+    if 'error' in msg:
+        raise reauth_errors.ReauthAPIError(msg['error']['message'])
+    return msg
+
+
+def GetUserPassword(text):
+    """Get password from user.
+
+    Override this function with a different logic if you are using this library
+    outside a CLI. Returns the password."""
+
+    return getpass.getpass(text)
+
+
+def InteractiveCheck():
+    """Check if we are in an interractive environment.
+
+    If the rapt token needs refreshing, the user needs to answer the
+    challenges.
+    If the user is not in an interractive environment, the challenges can not
+    be answered and we just wait for timeout for no reason."""
+
+    return sys.stdin.isatty()
+
+
+def GetPrintCallback():
+    """Get preferred output function."""
+
+    return sys.stderr.write
+
+
+class ReauthChallenge(object):
```

Comment:
Use `abc` to make sure that subclasses of this can not be instantiated without implementing the abstract methods.